man/man8/readahead.8
.TH readahead 8  "2020-08-20" "USER COMMANDS"
.SH NAME
readahead \- Show performance of read-ahead cache
.SH SYNOPSIS
.B readahead [-d DURATION]
.SH DESCRIPTION
The tool shows the performance of read-ahead caching on the system under a given load to investigate any
caching issues. It shows a count of unused pages in the cache and also prints a histogram showing how
long they have remained there.

This tool traces the \fB__do_page_cache_readahead()\fR kernel function to track entry and exit in the
readahead mechanism in the kernel and then uses \fB__page_cache_alloc()\fR and \fBmark_page_accessed()\fR
functions to calculate the age of the page in the cache as well as see how many are left unaccessed.

Since this uses BPF, only the root user can use this tool.
.SS NOTE ON KPROBES USAGE
Since the tool uses Kprobes, depending on your linux kernel's compilation, these functions may be inlined
and hence not available for Kprobes. To see whether you have the functions available, check \fBvmlinux\fR
source and binary to confirm whether inlining is happening or not. You can also check \fB/proc/kallsyms\fR
on the host and verify if the target functions are present there before using this.
.SH REQUIREMENTS
CONFIG_BPF, bcc
.SH OPTIONS
\-h
Print usage message
.TP
\-d DURATION
Trace the read-ahead caching system for DURATION seconds
.SH EXAMPLES
.TP
Trace for 30 seconds and show  histogram of page age (ms) in read-ahead cache along with unused page count:
#
.B readahead -d 30
.SH OVERHEAD
The kernel functions instrumented by this program could be high-frequency depending on the profile of the
application (for example sequential IO). We advise the users to measure and monitor the overhead before leaving
this turned on in production environments.
.SH SOURCE
This originated as a bpftrace tool from the book "BPF Performance Tools",
published by Addison Wesley (2019):
.IP
http://www.brendangregg.com/bpf-performance-tools-book.html
.PP
See the book for more documentation on this tool.
.PP
This version is in the BCC repository:
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Suchakra Sharma
.SH SEE ALSO
readahead(2), madvise(2)


debian/changelog
bcc (0.35.0-1) unstable; urgency=low

  * Support for kernel up to 6.14
  * New bcc tools: mptcp: enable mptcp for tcp traffic
  * tools/biosnoop: Fix biosnoop pattern option
  * Allow cmake run out of the source tree
  * Fix for test bpf_stack_id when running on custom image in yocto
  * libbpf-tools/map_helpers: Add bpf_map_lookup_and_delete_batch to dump_hash
  * libbpf-tools/biotop: Use dump_hash for map processing
  * uninstall: use execute_process() instead of exec_program()
  * clang: Fix pointer dereference on big-endian machines
  * ci: Upgrade to ubuntu 24.04
  * doc update, other bug fixes and example improvement.

bcc (0.34.0-1) unstable; urgency=low

  * Support for kernel up to 6.13
  * Bump cmake minimum version to 3.12
  * statsnoop: Display syscall name with -s
  * readahead: Fix incorrect page accessed count since kernel 5.16
  * libbpf-tools/opensnoop: Add new fields
  * libbpf-tools: hardirqs/softirqs: Fix logarithmic calculation issue
  * libbpf-tools/hardirqs: have better default display and add CPU column
  * libbpf-tools/klockstat: Better stack dump and summary info
  * libbpf-tools/sigsnoop: Support real-time signals and thread comm
  * libbpf-tools/statsnoop: Support more syscalls
  * libbpf-tools/memleak: Some fixes and better messages
  * tools/opensnoop: Add new fields and fix bad mode value
  * tools/profile: Prioritize using the cpu-cycles hardware event
  * tools/tcpdrop: Add support for dumping TCP drop reasons
  * Fix event name too long error in python source
  * doc update, other bug fixes and example improvement.

bcc (0.33.0-1) unstable; urgency=low

  * Support for kernel up to 6.12
  * Add new bcc tool numasched
  * syms: Initialize ModulePath::fd_ to invalid FD
  * libbpf-tools/memleak: Fix off-by-one error
  * libbpf-tools/slabratetop: Fix failed to create kprobe error
  * libbpf_tools/profile: Support PID namespace mapping
  * libbpf-tools/mountsnoop: Support fsopen,fsconfig,fsmount,move_mount syscalls
  * tools/oomkill: get application level stack trace
  * tools/profile: Add additional information to backtrace
  * tools/mountsnoop: Fix fsmount printing wrong flags
  * tools/compactsnoop: Add aarch64 support
  * doc update, other bug fixes and tools improvement.

bcc (0.32.0-1) unstable; urgency=low

  * Support for kernel up to 6.11.
  * bcc tool update: wakeuptime, readahead, shmsnoop, offcputime, cachestat, cachetop, hardirqs
  * libbpf tool update: futexctn, profile, readhead, softirqs, hardirqs
  * Multiple enhancements for memleak, better error path checking, adding mremap uprobe
  * Support get pid/tgid in pid namespaces (cpudist, profile)
  * multiple pid filtering support: profile, offcputime
  * detect whether elf binary is PIE even if the binary is marked as DYN
  * Fix several compilation issues with llvm20
  * doc update, other bug fixes and tools improvement.

bcc (0.31.0-1) unstable; urgency=low

  * Support for kernel up to 6.9.
  * Add support for bcachefs to fsdist and fsslower tools
  * libbpf tool update: memleak, syncsnoop, numamove, syscount, vfsstat, tcptop, capable, syncsnoop, sigsnoop, etc.
  * bcc tool update: biolatency, biosnoop, biotop, vfsstat, kvmexit, sslsniff, swapin, etc.
  * build: Remove llvm-dev dependency from libbcc
  * build: Remove dependency on LLVM header from libbcc packages
  * usdt: Fix bare register dereference on aarch64
  * Extend `bcc_proc` API which allows to limit search to specific pid
  * Fix several flaky tests.
  * doc update, other bug fixes and tools improvement.

 -- Yonghong Song <ys114321@gmail.com>  Sat, 27 Jul 2024 17:00:00 +0000

bcc (0.30.0-1) unstable; urgency=low

  * Support for kernel up to 6.8.
  * Set minimum supported llvm version to 12, and add llvm17 test.
  * Add workqueue latency observation tool.
  * libbpf tool update: f2fsslower, opensnoop, futexctn, bindsnoop, ksnoop, klockstat, offcputime, etc.
  * bcc tool update: memleak, ttysnoop, bashreadline, tcpdrop, execsnoop, etc.
  * allow more flexible perf event options with new perf_custom_event_open() python API.
  * Fix userspace stack unwinding on powerpc.
  * add bpf_prog_test_run_opts() python API.
  * several deb package related changes.
  * Fix btf_type_tag issue with llvm 15.
  * Fix several flaky tests.
  * classify tools into different sub-categories.
  * doc update, other bug fixes and tools improvement.

 -- Yonghong Song <ys114321@gmail.com>  Thu, 24 Mar 2024 17:00:00 +0000

bcc (0.29.1-1) unstable; urgency=low

  * Fix Ubuntu 22.04 deb build
  * Fix Ubuntu 22.04 Docker image build

 -- Adrian Vladu <avladu@cloudbasesolutions.com>  Thu, 14 Dec 2023 17:00:00 +0000

bcc (0.29.0-1) unstable; urgency=low

  * Support for kernel up to 6.6.
  * new bcc tools: rdmaucma
  * new libbpf tools: f2fs, futexctn
  * bcc tool update: tcpstates, statsnoop, runqlat, bio tools, tcptop, slabratetop, tcprtt, etc.
  * libbpf tool update: tcprtt, tcppktlat, bio tools, execsnoop, bindsnoop, exitsnoop, etc.
  * s390x support for libbpf-tools
  * examples for perf/ipc
  * expose pid parameter in bpf_open_perf_event
  * allow for installing python as a non-system package
  * ci improvement: deprecate ubuntu 18.04 and allow multiple llvm versions
  * ci improvement: reitre fedora 34/26 and use fedora 38.
  * fix misaligned pointer accesses in some ringbuf using libbpf-tools
  * some new  enhancement for powerpc and riscv.
  * consolidate tools into different categories: filesystem/storage, networking, cpu and scheduling, etc.
  * doc update, other bug fixes and tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Thu, 6 Dec 2023 17:00:00 +0000

bcc (0.28.0-1) unstable; urgency=low

  * Support for kernel up to 6.3.
  * new libbpf tool: tcppktlat.
  * bcc tool updates: funcslower, wakeuptime, profile, offcputime, deadlock, funccount, argdist, kvmexit, runqlen and cpuunclaimed.
  * libbpf tool update: memleak, tcprtt, tcpconnlat, funclatency, syscount, cpufreq, biosnoop.
  * support ringbuf_query for bcc tools.

tools/readahead.py
#!/usr/bin/env python
# @lint-avoid-python-3-compatibility-imports
#
# readahead     Show performance of read-ahead cache
#               For Linux, uses BCC, eBPF
#
# Copyright (c) 2020 Suchakra Sharma <mail@suchakra.in>
# Licensed under the Apache License, Version 2.0 (the "License")
# This was originally created for the BPF Performance Tools book
# published by Addison Wesley. ISBN-13: 9780136554820
# When copying or porting, include this comment.
#
# 20-Aug-2020   Suchakra Sharma     Ported from bpftrace to BCC
# 17-Sep-2021   Hengqi Chen         Migrated to kfunc
# 30-Jan-2023   Rong Tao            Support more kfunc/kprobe, introduce folio

from __future__ import print_function
from bcc import BPF
from time import sleep
import ctypes as ct
import argparse

# arguments
examples = """examples:
    ./readahead -d 20       # monitor for 20 seconds and generate stats
"""

parser = argparse.ArgumentParser(
    description="Monitor performance of read ahead cache",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-d", "--duration", type=int,
    help="total duration to monitor for, in seconds")
args = parser.parse_args()
if not args.duration:
    args.duration = 99999999

# BPF program
bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/mm_types.h>
#include <linux/mm.h>

BPF_HASH(flag, u32, u8);            // used to track if we are in do_page_cache_readahead()
BPF_HASH(birth, struct page*, u64); // used to track timestamps of cache alloc'ed page
BPF_ARRAY(pages);                   // increment/decrement readahead pages
BPF_HISTOGRAM(dist);
"""

bpf_text_kprobe = """
int entry__do_page_cache_readahead(struct pt_regs *ctx) {
    u32 pid;
    u8 one = 1;
    pid = bpf_get_current_pid_tgid();
    flag.update(&pid, &one);
    return 0;
}

int exit__do_page_cache_readahead(struct pt_regs *ctx) {
    u32 pid;
    u8 zero = 0;
    pid = bpf_get_current_pid_tgid();
    flag.update(&pid, &zero);
    return 0;
}

int exit__page_cache_alloc(struct pt_regs *ctx) {
    u32 pid;
    u64 ts;
    struct page *retval = (struct page*) GET_RETVAL_PAGE;
    u32 zero = 0; // static key for accessing pages[0]
    pid = bpf_get_current_pid_tgid();
    u8 *f = flag.lookup(&pid);
    if (f != NULL && *f == 1) {
        ts = bpf_ktime_get_ns();
        birth.update(&retval, &ts);
        pages.atomic_increment(zero);
    }
    return 0;
}

int entry_mark_page_accessed(struct pt_regs *ctx) {
    u64 ts, delta;
    struct page *arg0 = GET_ARG1_PAGE;
    u32 zero = 0; // static key for accessing pages[0]
    u64 *bts = birth.lookup(&arg0);
    if (bts != NULL) {
        delta = bpf_ktime_get_ns() - *bts;
        dist.atomic_increment(bpf_log2l(delta/1000000));
        pages.atomic_increment(zero, -1);
        birth.delete(&arg0); // remove the entry from hashmap
    }
    return 0;
}
"""

bpf_text_kfunc_cache_readahead = """
KFUNC_PROBE(RA_FUNC)
{
    u32 pid = bpf_get_current_pid_tgid();
    u8 one = 1;

    flag.update(&pid, &one);
    return 0;
}

KRETFUNC_PROBE(RA_FUNC)
{
    u32 pid = bpf_get_current_pid_tgid();
    u8 zero = 0;

    flag.update(&pid, &zero);
    return 0;
}
"""

bpf_text_kfunc_mark_accessed_template = """
KFUNC_PROBE(MA_FUNC_NAME, MA_ARG_TYPE arg0)
{
    u64 ts, delta;
    u32 zero = 0; // static key for accessing pages[0]
    struct page *page = GET_PAGE_PTR_FROM_ARG0;
    u64 *bts = birth.lookup(&page);

    if (bts != NULL) {
        delta = bpf_ktime_get_ns() - *bts;
        dist.atomic_increment(bpf_log2l(delta/1000000));
        pages.atomic_increment(zero, -1);
        birth.delete(&page); // remove the entry from hashmap
    }
    return 0;
}
"""

bpf_text_kfunc_cache_alloc_ret_page = """
KRETFUNC_PROBE(__page_cache_alloc, gfp_t gfp, struct page *retval)
{
    u64 ts;
    u32 zero = 0; // static key for accessing pages[0]
    u32 pid = bpf_get_current_pid_tgid();
    u8 *f = flag.lookup(&pid);

    if (f != NULL && *f == 1) {
        ts = bpf_ktime_get_ns();
        birth.update(&retval, &ts);
        pages.atomic_increment(zero);
    }
    return 0;
}
"""

bpf_text_kfunc_cache_alloc_ret_folio = """
KRETFUNC_PROBE(filemap_alloc_folio, gfp_t gfp, unsigned int order,
    struct folio *retval)
"""
# In kernel commit b951aaff5035 ("mm: enable page allocation tagging"), add
# _noprof suffix to filemap_alloc_folio.
bpf_text_kfunc_cache_alloc_ret_folio_noprof = """
KRETFUNC_PROBE(filemap_alloc_folio_noprof, gfp_t gfp, unsigned int order,
    struct folio *retval)
"""
bpf_text_kfunc_cache_alloc_ret_folio_func_body = """
{
    u64 ts;
    u32 zero = 0; // static key for accessing pages[0]
    u32 pid = bpf_get_current_pid_tgid();
    u8 *f = flag.lookup(&pid);
    struct page *page = folio_page(retval, 0);

    if (f != NULL && *f == 1) {
        ts = bpf_ktime_get_ns();
        birth.update(&page, &ts);
        pages.atomic_increment(zero);
    }
    return 0;
}
"""

if BPF.support_kfunc():
    if BPF.get_kprobe_functions(b"__do_page_cache_readahead"):
        ra_func = "__do_page_cache_readahead"
    elif BPF.get_kprobe_functions(b"do_page_cache_ra"):
        ra_func = "do_page_cache_ra"
    elif BPF.get_kprobe_functions(b"page_cache_ra_order"):
        ra_func = "page_cache_ra_order"
    else:
        print("Not found any kfunc for page cache readahead.")
        exit()
    bpf_text += bpf_text_kfunc_cache_readahead.replace("RA_FUNC", ra_func)
    if BPF.get_kprobe_functions(b"__page_cache_alloc"):
        bpf_text += bpf_text_kfunc_cache_alloc_ret_page
    else:
        if BPF.get_kprobe_functions(b"filemap_alloc_folio"):
            bpf_text += bpf_text_kfunc_cache_alloc_ret_folio
        elif BPF.get_kprobe_functions(b"filemap_alloc_folio_noprof"):
            bpf_text += bpf_text_kfunc_cache_alloc_ret_folio_noprof
        else:
            print("ERROR: No cache alloc function found. Exiting.")
            exit()
        bpf_text += bpf_text_kfunc_cache_alloc_ret_folio_func_body
    if BPF.get_kprobe_functions(b"folio_mark_accessed"):
        ma_func_name = "folio_mark_accessed"
        ma_arg_type = "struct folio *"
        get_page_ptr_code = "folio_page(arg0, 0)"
        bpf_text_kfunc_mark_accessed = bpf_text_kfunc_mark_accessed_template \
            .replace("MA_FUNC_NAME", ma_func_name) \
            .replace("MA_ARG_TYPE", ma_arg_type) \
            .replace("GET_PAGE_PTR_FROM_ARG0", get_page_ptr_code)
    elif BPF.get_kprobe_functions(b"mark_page_accessed"):
        ma_func_name = "mark_page_accessed"
        ma_arg_type = "struct page *"
        get_page_ptr_code = "arg0"
        bpf_text_kfunc_mark_accessed = bpf_text_kfunc_mark_accessed_template \
            .replace("MA_FUNC_NAME", ma_func_name) \
            .replace("MA_ARG_TYPE", ma_arg_type) \
            .replace("GET_PAGE_PTR_FROM_ARG0", get_page_ptr_code)
    else:
        print("Not found any kfunc for page cache mark accessed.")
        exit()
    bpf_text += bpf_text_kfunc_mark_accessed
    b = BPF(text=bpf_text)
else:
    bpf_text += bpf_text_kprobe
    if BPF.get_kprobe_functions(b"__do_page_cache_readahead"):
        ra_event = "__do_page_cache_readahead"
    elif BPF.get_kprobe_functions(b"do_page_cache_ra"):
        ra_event = "do_page_cache_ra"
    elif BPF.get_kprobe_functions(b"page_cache_ra_order"):
        ra_event = "page_cache_ra_order"
    else:
        print("Not found any kprobe for page cache readahead.")
        exit()
    if BPF.get_kprobe_functions(b"__page_cache_alloc"):
        cache_func = "__page_cache_alloc"
        bpf_text = bpf_text.replace('GET_RETVAL_PAGE', 'PT_REGS_RC(ctx)')
    else:
        if BPF.get_kprobe_functions(b"filemap_alloc_folio"):
            cache_func = "filemap_alloc_folio"
        elif BPF.get_kprobe_functions(b"filemap_alloc_folio_noprof"):
            cache_func = "filemap_alloc_folio_noprof"
        else:
            print("ERROR: No cache alloc function found. Exiting.")
            exit()
        bpf_text = bpf_text.replace('GET_RETVAL_PAGE', 'folio_page((struct folio *)PT_REGS_RC(ctx), 0)')
    if BPF.get_kprobe_functions(b"folio_mark_accessed"):
        ma_event = "folio_mark_accessed"
        bpf_text = bpf_text.replace('GET_ARG1_PAGE', 'folio_page((struct folio *)PT_REGS_PARM1(ctx), 0)')
    elif BPF.get_kprobe_functions(b"mark_page_accessed"):
        ma_event = "mark_page_accessed"
        bpf_text = bpf_text.replace('GET_ARG1_PAGE', '(struct page *)PT_REGS_PARM1(ctx)')
    else:
        print("Not found any kprobe for page cache mark accessed.")
        exit()

    b = BPF(text=bpf_text)
    b.attach_kprobe(event=ra_event, fn_name="entry__do_page_cache_readahead")
    b.attach_kretprobe(event=ra_event, fn_name="exit__do_page_cache_readahead")
    b.attach_kretprobe(event=cache_func, fn_name="exit__page_cache_alloc")
    b.attach_kprobe(event=ma_event, fn_name="entry_mark_page_accessed")

# header
print("Tracing... Hit Ctrl-C to end.")

# print
def print_stats():
    print()
    print("Read-ahead unused pages: %d" % (b["pages"][ct.c_ulong(0)].value))
    print("Histogram of read-ahead used page age (ms):")
    print("")
    b["dist"].print_log2_hist("age (ms)")
    b["dist"].clear()
    b["pages"].clear()

while True:
    try:
        sleep(args.duration)
        print_stats()
    except KeyboardInterrupt:
        print_stats()
        break
